x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -106,6 +108,11 @@ public void beforeStop() {
                 shutdown();
             }
         });
+        clusterService.addListener((event) -> {
+            if (event.metadataChanged()) {
+                isResetMode = MlMetadata.getMlMetadata(event.state()).isResetMode();
```

Comment:
I think there is a nasty edge case with this.

The scenario is that a reset starts and finishes while this class is sleeping between retries.  Then when it wakes up it writes documents that were retrying from before the refresh.

One way around this that I can think of is to have an incrementing count of how many times reset mode has toggled.  Any scheduled retries that wake up check if that count is different to what it was when they were originally scheduled, and if there's a difference they do nothing.

Another way might be to cancel any scheduled retries when this reset mode flag changes value.

Reply:
> Another way might be to cancel any scheduled retries when this reset mode flag changes value.

This might be the easiest way as we already keep track of which retryable actions are inflight. Will just have to separate out the search and the bulk inflight actions.